src/cockpit/protocol.py

```diff
@@ -40,10 +38,11 @@ class CockpitProblem(Exception):
     It is usually thrown in response to some violation of expected protocol
     when parsing messages, connecting to a peer, or opening a channel.
     """
-    def __init__(self, problem: str, **kwargs):
-        super().__init__(kwargs.get('message') or problem)
+    def __init__(self, problem: str, _msg: 'JsonObject | None' = None, **kwargs: JsonDocument) -> None:
+        self.attrs = create_object(_msg, kwargs)
+        self.attrs['problem'] = problem
         self.problem = problem
-        self.kwargs = kwargs
```

Comment:
I didn't bother extensively applying this stuff to code paths that are exclusively related to channels since in those cases (with the sole exception of `pong` messages) we're always generating new messages rather than passing data through.

`ChannelError` is a direct derivative of `Exception` and has nothing at all to do with `CockpitProblem`.  We could actually consider merging those, I suppose...

`RoutingError` is the same.  My oh my we have a lot of exception types that are all the same...

I could do some additional consolidation here if you think that it would be meaningful.

Reply:
Looking into this deeper, it's not clear that this would be an improvement, in fact.  There are zero cases in the code of wanting to pass messages straight-through on this path (ie: we never raise ChannelError with **kwargs from somewhere else) nor do we call Channel.close() with `**kwargs` from somewhere that we don't want to apply the processing rules to (and almost all cases of that are the handling of ChannelError you linked above).

Reply:
The issue here is that there are many direct users of `Channel.close()` which give kwargs.

I suppose we could modify `Channel.close() to take the `_msg, **kwargs` pair that we use elsewhere, and call `create_object()` internally to create the `_close_args` dict.  We could then pass the `_close_args` dict directly to the `_msg` parameter on `shutdown_endpoint()` instead of using a `**` spread there.  That would be a minor improvement.

Another possibility would be to try to port all existing users of `Channel.close()` with non-empty kwargs list to raise `ChannelError` instead.  There's one problematic case which is how the filesystem backend returns tags in closes messages, including in non-error cases, but we could fix that by just passing a dictionary directly instead of using kwargs.  I'll give this a try.

Reply:
OK, I see, these are different custom error classes. I still think it's confusing that some of our errors have `exc.kwargs` and some have `exc.attrs` while they both have exactly the same purpose and grammar. I don't think we should unify the error *classes* (as it's useful to have distinct types for them), just the vocabulary between them. However, it seems that you don't consider them having the same purpose, let's discuss in the next thread.

Reply:
OK.  I've unified all three classes.  PeerError is also somewhat redundant, but it's taken care of in my peer rework branch.  I'm pretty happy about this cleanup, thanks. :)